src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -496,14 +513,113 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
+        //
+        // Document for the snippet is included here in this code block, since
+        // shader length affects compilation time  significantly on some
+        // backends.
+        //
+        // Snippet with documentation:
+        //   const kExponentBias = 1022;
+        //
+        //   // Detect if the value is zero or subnormal, so that FTZ behaviour
+        //   // can occur
+        //   const subnormal_or_zero : bool = (${expr} <= ${kValue.f64.subnormal.positive.max}) && (${expr} >= ${kValue.f64.subnormal.negative.min});
+        //
+        //   // MSB of the high_u32 is 1 if the value is negative, otherwise 0
+        //   // Extract the sign bit early, so that abs() can be used with
+        //   // frexp() to simplify a bunch of logic later
+        //   const sign_bit : u32 = select(0, 0x80000000, ${expr} < 0);
+        //
+        //   // Use frexp() to obtain the exponent and fractional parts, and
+        //   // then perform FTZ if needed
+        //   const f = frexp(abs(${expr}));
+        //   const f_fract = select(f.fract, 0, subnormal_or_zero);
+        //   const f_exp = select(f.exp, -kExponentBias, subnormal_or_zero);
+        //
+        //   // The exponent is stored in bits [20..31] of high_u32, special
+        //   // case for zero.
+        //   const exponent_bits : u32 = (f_exp + kExponentBias) << 20;
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, masking off
+        //   // unneeded bits.
+        //   const high_mantissa_bits : u32 = (u32(f_fract * (1 << 20))) & 0x000fffff;
+        //
+        //   // Subtract out the high bits from the mantissa.
+        //   // const low_mantissa = f_fract - floor(f_fract * (1<< 21)) / (1 << 21);
+        //
+        //   // Extract the low 32 bits of the mantissa as an u32
+        //   // The calculated lower mantissa may or may not have a leading 1 in
+        //   // for the 33rd bit that needs to be masked.
+        //   // Unfortunately the value at this point is an AbstractFloat, which
+        //   // cannot be directly converted to an AbstractInt, so there is no
+        //   // simple way to just mask the bits that are needed without
+        //   // concretizing first.
+        //   // If that bit is present the u32 cast will fail due to being out
+        //   // of range. If the bit isn't present, but removed without checking
+        //   // for its presences, the result will be negative and the wrong
+        //   // magnitude.
+        //   // So when converting to u32, the value needs to be selected on to
+        //   // remove that bit iff it is present.
+        //   const low_mantissa_bits = (low_mantissa * (1 << 52));
+        //
+        //   // Pack the result into 2 u32s, adjusting the low mantissa if
+        //   // needed
+        //   outputs[${i}].value.x = u32(select(low_mantissa_bits, low_mantissa_bits - 0x100000000, low_mantissa_bits >= 0x100000000));
```

Comment:
The only reason I can think that there is a 1 in the 33rd position is if there is a numerical error in the computation somewhere. Maybe it's in the division `1 / (1<<21)`


I recoded the algorithm using ldexp instead and it's much more straightforward, I think.  

This uses C++17 so I can use hex floats.  (And I assume `unsigned` is the same as `uint32_t`, which is common these days.)

```cpp
#include <iostream>
#include <iomanip>
#include <cmath>

#define SHOW(X)  std::cout << std::setw(15) << " " #X << "  " << std::hex << std::hexfloat << (X) << std::endl

int main() {

  const double orig = 0x1.921fb54442d18p+1; // This is PI
  int n;
  const auto f = frexp(orig,&n);
  SHOW(f);

  const auto highf = ldexp(f,21); // 21 is the 20 bits we want plus the implicit leading 1
  SHOW(highf);
  const auto high = unsigned(highf);
  SHOW(high);

  const auto lowf = f - ldexp(floor(highf),-21);
  SHOW(lowf);
  const auto low = unsigned(ldexp(lowf,53));   // 53 = 32 + 21
  SHOW(low);
}
```
And here is the output:

```
$ ./a
              f  0x1.921fb54442d18p-1
          highf  0x1.921fb54442d18p+20
           high  1921fb
           lowf  0x1.5110b46p-23
            low  54442d18
```

Note that 'high' still has the 21st bit set, so you do have to mask it off.

They key is we recover exactly the right bits, unchanged.

Reply:
Done